compiler/src/dotty/tools/dotc/transform/init/Semantic.scala

```diff
@@ -231,62 +294,43 @@ class Semantic {
           Result(value2, errors)
       }
 
-    def call(meth: Symbol, superType: Type, source: Tree): Contextual[Result] =
+    def call(meth: Symbol, superType: Type, source: Tree, needResolve: Boolean = true): Contextual[Result] =
       value match {
         case Hot  =>
           Result(Hot, Errors.empty)
 
         case Cold =>
-          val error = CallCold(meth, source, trace)
+          val error = CallCold(meth, source, trace.toVector)
           Result(Hot, error :: Nil)
 
-        case thisRef: ThisRef =>
+        case addr: Addr =>
           val target =
-            if superType.exists then
-              resolveSuper(thisRef.klass, superType, meth)
+            if !needResolve then
+              meth
+            else if superType.exists then
+              resolveSuper(addr.klass, superType, meth)
             else
-              resolve(thisRef.klass, meth)
+              resolve(addr.klass, meth)
           if target.isOneOf(Flags.Method | Flags.Lazy) then
             if target.hasSource then
+              val cls = target.owner.enclosingClass.asClass
```

Comment:
This changed from `target.owner.asClass` to `target.owner.enclosingClass.asClass`. Is that intended?

Reply:
Yes, it's intentional -- here we reuse the code for calling local methods.

The related code change can be found below for `case id: Ident =>`.